File: leetcode/HouseRobber.java
public class HouseRobber {
  public int rob(int[] nums) {
    // dp[i] the max value achieved by 0..i-1 houses.
    int[] dp = new int[nums.length + 1];
    for (int i = 1; i <= nums.length; ++i) {
      if (i >= 2) {
        dp[i] = Math.max(dp[i-1], nums[i-1] + dp[i-2]);
      } else {
        dp[i] = Math.max(dp[i-1], nums[i-1]);
      }
    }
    return dp[nums.length];
  }
}

File: leetcode/HouseRobber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        vector<vector<int>> dp(n, vector<int>(n, 0));
        for (int i = 0; i < n; ++i) {
            dp[i][i] = nums[i];
        }
        for (int step = 2; step <= n; ++step) {
            for (int i = 0; i + step <= n; ++i) {
                int j = i + step - 1;
                int maxValue = 0;
                for (int k = i; k <= j; ++k) {
                    int value = nums[k];
                    if (k - 2 >= i) {
                        value += dp[i][k-2];
                    }
                    if (k + 2 <= j) {
                        value += dp[k + 2][j];
                    }
                    maxValue = max(maxValue, value);
                }
                dp[i][j] = maxValue;
            }
        }
        return dp[0][n-1];
    }
};
